#162701189 Fix retrieve specific article

Darius-Ndubi opened the pull request:
## What does this PR do?
 - Fixes get specific article route

## Description of Task to be completed?
- Adds route to get a specific article
-  Adds test on retrieving a specific article

## How should this be manually tested?

- Clone the repo and cd into the project folder
- Create a virtual environment by running `$ python3 -m venv <name_of_environment>`
- Activate the virtual environment by running `source <name_of_environment>/bin/activate`
- Run `$ pip install -r requirements.txt`
- Then add the following environment variables to your computer.

  - `$ export DB_NAME='choose_a_dbname'`
  - `$ export DB_USER='db_username'`
  - `$ export DB_PASSWORD='db_role_password'`
  - `$ export DB_HOST='localhost or preffered host'`
  - `$ export DB_PORT='server_port_number'`
  - `$ export APP_DEBUG='True' `
  - `$ export SOCIAL_AUTH_FACEBOOK_KEY='<facebook_app_id>'`
  - `$ export SOCIAL_AUTH_FACEBOOK_SECRET='<facebook_app_secret>'`
  - `$ export SOCIAL_AUTH_TWITTER_KEY='<twitter_api_key>'`
  - `$ export SOCIAL_AUTH_TWITTER_SECRET='<twitter_api_secret>'`
  - `$ export SOCIAL_AUTH_GOOGLE_OAUTH2_KEY='<google_client_id>'`
  - `$ export SOCIAL_AUTH_GOOGLE_OAUTH2_SECRET='<google_client_secret>'`

- Change the current branch to bg-get-specific-article-162701189
-  Run `python3 manage.py makemigrations`
-  Run `python3 manage.py migrate`
-  Run `python3 manage.py test`

## What are the relevant pivotal tracker stories?
[#162701189](https://www.pivotaltracker.com/story/show/162701189)

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

## Screenshots 
### Get existing article
<img width="1280" alt="screenshot 2018-12-17 at 19 42 06" src="https://user-images.githubusercontent.com/16937341/50101551-ffd13880-0233-11e9-8d19-f3b5184e800f.png">

### Get non existing article
<img width="1280" alt="screenshot 2018-12-17 at 19 42 24" src="https://user-images.githubusercontent.com/16937341/50101595-15466280-0234-11e9-9d45-86761d52f759.png">

bettblake08 commented on authors/apps/articles/views.py:
```diff
@@ -43,6 +43,25 @@ class ArticleView(RetrieveUpdateDestroyAPIView):
+        # When article is not found, Give the user a description
+        # of what happened. If the article is found return it back to user
+
+        if not Article.objects.filter(slug=slug):
```
Have the article query in a try-catch clause to catch the not found error. This reduces the number of queries to the database.

Darius-Ndubi replied:
Okay @bettblake08 let me switch to that.

Darius-Ndubi replied:
Switched out to try except, Thanks for the feedback. Please check on the update.

Bryan-Cee commented on authors/apps/articles/views.py:
```diff
@@ -43,6 +43,24 @@ class ArticleView(RetrieveUpdateDestroyAPIView):
+        except Article.DoesNotExist:
+            raise exceptions.NotFound({
+                "message": "Article was not found"})
+        return Response(
```
I believe it would be more easier to use serializer and JSON render to handle the return values. This will make it easier to get the authors' profile data which should also be included in the return response, which currently is not.

denomash reviewed (approved):
👍

mutoromike commented on authors/apps/articles/views/comments.py:
```diff
@@ -260,34 +172,40 @@ def put(self, request, *args, **kwargs):
+            error_response = {
+                "response": {
                     "user": {
                         "user": RESPONSE['not_found'].format(data="User")
```
I believe that this should be `unauthorized` and not a `not_found`

Darius-Ndubi replied:
@mutoromike  we have updated it to use forbidden for a message we have in response.py

mutoromike commented on authors/apps/articles/views/comments.py:
```diff
@@ -260,34 +172,40 @@ def put(self, request, *args, **kwargs):
+            error_response = {
+                "response": {
                     "errors": {
                         "text": RESPONSE['empty_field'].format('text')
```
Use descriptive messages like `empty fields not allowed`

mutoromike reviewed (changes requested):
Kindly address the changes I've requested.

Darius-Ndubi committed:
Fix(get article) : Retrieve specific article

 - Add get specific article
 - Add renderer for article response
 - Add serializer for article
 - Add test to get specific article
 - Split views into articles and comments

[Delivers #162701189]
